src/main/java/de/learnlib/ralib/tools/IOSimulator.java

```diff
@@ -303,10 +303,9 @@ public void run() {
             SimpleProfiler.stop(__LEARN__);
             SimpleProfiler.start(__EQ__);
             DefaultQuery<PSymbolInstance, Boolean> ce  = null;
-            DefaultQuery<PSymbolInstance, Boolean> origCe  = null;
 
             if (useEqTest) {
-                ce = this.eqTest.findCounterExample(hyp, null);
+                ce = this.eqTest.findCounterExample(hyp, new ArrayList<>());
```

Comment:
I've looked a bit more on this.  Here are two things I've found:
1. In the entire code base of ralib, we pass `null` for this argument.  I think this shows that either this argument is not needed in practice (and LearnLib's interface has been over-engineered; perhaps its interface should only specify the 1-argument version of this method), or we do not use the functionality of this method properly.
2. If I am not mistaken, this particular call which is flagged in this comment refers to the definition of this function in [IOEquivalenceTest.java](https://github.com/LearnLib/ralib/blob/main/src/main/java/de/learnlib/ralib/equivalence/IOEquivalenceTest.java#L230).  That argument is clearly unused there; we may as well pass anything there...

Reply:
I think it is over-engineered in LearnLib. @mtf90: can you suggest something here?

Reply:
Just to pick up on @kostis' previous comment: both versions are problematic. The parameter is meant to determine which input symbols should be used for finding counterexamples. So `null` would get flagged by any null-analysis and `new ArrayList<>()` could (if implemented accordingly) at most return epsilon as a counterexample. The reason why this works in this case is (probably) because you only use the `IOEquivalenceTest` and `IORandomWalk` which ignore this parameter and use an internal field for the inputs.

As for the question whether this is over-engineered, I think it has its benefits. Passing the eligible input symbols as a parameter allows the equivalence oracle to not have to store the inputs. Especially for LearnLib features such as `Resumable` or `SupportsGrowingAlphabet` (which allow you to snapshot learning states and expand in different input-directions) this is quite useful because it reduces data dependencies. This is currently already a hassle with caches which do store the referenced alphabet and therefore need to be updated as well when new input symbols are introduced. By removing this parameter this management would also be necessary for equivalence oracles.

I can see your point that it adds somewhat of an overhead in the learning loop, but it also reduces overhead when implementing equivalence oracles. So for right now, I would suggest to keep the interface as is and try to simplify the equivalence oracles if possible.